develop/blockchains/custom-blockchains/benchmarking.md

```diff
@@ -1,4 +1,191 @@
+Including or excluding transactions based on available resources ensures that the runtime can continue to produce and import blocks without service interruptions. For example, suppose you have a function call that requires particularly intensive computation. In that case, executing the call might exceed the maximum time allowed for producing or importing a block, disrupting the block handling process or stopping blockchain progress altogether. Benchmarking helps you validate that the execution time required for different functions is within reasonable boundaries.
+
+Similarly, a malicious user might attempt to disrupt network service by repeatedly executing a function call that requires intensive computation or doesn't accurately reflect the necessary computation. If the cost of executing a function call doesn't accurately reflect the computation involved, there's no incentive to deter a malicious user from attacking the network. 
+
+!!!info "Benchmarking helps with predictable fees and computational outcomes."
+ Because benchmarking helps you evaluate the weight associated with executing transactions, it also helps you to determine appropriate transaction fees. Based on your benchmarks, you can set fees representing the resources consumed by executing specific calls on the blockchain.
+
+### Benchmarking and Weight 
+
+Polkadot SDK-based chains use weight to represent the time it takes to execute the transactions in a block. The time required to execute any particular call in a transaction depends on several factors, including the following:
+
+- Computational complexity
+- Storage complexity
+- Database read and write operations required
+- Hardware used
+  
+To calculate an appropriate weight for a transaction, you can use benchmark parameters to measure the time it takes to execute the function calls on different hardware, using different values and repeating them multiple times. You can then use the results of the benchmarking tests to establish an approximate worst-case weight to represent the resources required to execute each function call and each code path. Fees are then based on the worst-case weight. If the actual call performs better than the worst case, the weight is adjusted, and any excess fees can be returned.
+
+Because weight is a generic unit of measurement based on computation time for a specific physical machine, the weight of any function can change based on the specific hardware used for benchmarking.
+
+By modeling the expected weight of each runtime function, the blockchain can calculate the number of transactions or system-level calls it can execute within a certain period of time.
+
+Within FRAME, each function call that is dispatched must have a `#[weight]` annotation that can return the expected weight for the worst-case scenario execution of that function given its inputs. The benchmarking framework automatically generates a file with those formulas for you, which you can use in that annotation within your pallet.
```

Comment:
Here you can add a code example of a real extrinsic with its actual macro annotation